documentation-support/src/main/java/tech/picnic/errorprone/documentation/BugPatternTestExtractor.java

```diff
@@ -0,0 +1,180 @@
+            .named("addSourceLines");
+    private static final Matcher<ExpressionTree> REPLACEMENT_INPUT =
+        instanceMethod()
+            .onDescendantOf("com.google.errorprone.BugCheckerRefactoringTestHelper")
+            .named("addInputLines");
+    private static final Matcher<ExpressionTree> REPLACEMENT_OUTPUT =
+        instanceMethod()
+            .onDescendantOf("com.google.errorprone.BugCheckerRefactoringTestHelper.ExpectOutput")
+            .named("addOutputLines");
+
+    private final VisitorState state;
+    private final List<String> identificationTests = new ArrayList<>();
+    private final List<BugPatternReplacementTestDocumentation> replacementTests = new ArrayList<>();
+
+    @Var private String replacementOutputLines = "";
+
+    CollectBugPatternTests(VisitorState state) {
+      this.state = state;
+    }
+
+    public ImmutableList<String> getIdentificationTests() {
+      return ImmutableList.copyOf(identificationTests);
+    }
+
+    public ImmutableList<BugPatternReplacementTestDocumentation> getReplacementTests() {
+      return ImmutableList.copyOf(replacementTests);
+    }
+
+    @Override
+    public @Nullable Void visitMethodInvocation(MethodInvocationTree node, @Nullable Void unused) {
+      if (IDENTIFICATION_SOURCE_LINES.matches(node, state)) {
+        identificationTests.add(getSourceLines(node));
+      } else if (REPLACEMENT_INPUT.matches(node, state)) {
+        /* The visitor starts with `addOutputLines` and in the next visit it will go over the `addInputLines`. */
+        replacementTests.add(
+            BugPatternReplacementTestDocumentation.create(
+                getSourceLines(node), replacementOutputLines));
+      } else if (REPLACEMENT_OUTPUT.matches(node, state)) {
+        replacementOutputLines = getSourceLines(node);
+      }
+      return super.visitMethodInvocation(node, unused);
+    }
+
+    // XXX: Duplicate from `ErrorProneTestSourceFormat`, should we move this to `SourceCode` util?
+    private static String getSourceLines(MethodInvocationTree tree) {
+      List<? extends ExpressionTree> sourceLines =
+          tree.getArguments().subList(1, tree.getArguments().size());
+      StringBuilder source = new StringBuilder();
+
+      for (ExpressionTree sourceLine : sourceLines) {
+        Object value = ASTHelpers.constValue(sourceLine);
+        if (value == null) {
+          return "";
+        }
+        source.append(value).append('\n');
+      }
+
+      return source.toString();
+    }
+  }
+
+  @AutoValue
+  abstract static class BugPatternTestDocumentation {
+    abstract String name();
+
+    abstract ImmutableList<String> identificationTests();
+
+    abstract ImmutableList<BugPatternReplacementTestDocumentation> replacementTests();
+  }
+
+  @AutoValue
+  abstract static class BugPatternReplacementTestDocumentation {
+    static BugPatternReplacementTestDocumentation create(String sourceLines, String outputLines) {
```

Comment:
Couldn't find a way around this, the way we instantiate the `BugPatternTestDocumentation` earlier on wasn't possible with this class...